### PR TITLE
Raven db 6978

### PR DESCRIPTION
**Pull request by mobilunity-user**
Unit tests will be later. 
Library wasn't tested - PR is done just for your review

### CODE REVIEW

File: src/Database/Commands/DeleteDatabaseCommand.ts [hunk 1]
```diff
@@ -4,21 +4,21 @@ import {RequestMethods} from "../../Http/Request/RequestMethod";
...
 
 export class DeleteDatabaseCommand extends RavenCommand {
-    protected databaseId?: string;
-    protected hardDelete: boolean = false;
+  protected databaseId?: string;
+  protected hardDelete: boolean = false;
 
-    constructor(databaseId: string, hardDelete: boolean = false) {
-        super('', RequestMethods.Delete);
+  constructor(databaseId: string, hardDelete: boolean = false) {
```
Comment (gregolsky) [inline src/Database/Commands/DeleteDatabaseCommand.ts:10]: It's `databaseName` here and `fromNode` parameter is missing. Please see https://github.com/ravendb/ravendb/blob/v4.0/src/Raven.Client/Server/Operations/DeleteDatabaseOperation%20.cs
(command class is there on the bottom)

File: src/Database/Operations/OperationExecutor.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  protected requestExecutor: IRequestExecutor;
+
+  protected abstract requestExecutorFactory(): IRequestExecutor;  
+
+  constructor(store: IDocumentStore) {
+    this.store = store;
+    this.requestExecutor = this.requestExecutorFactory();
+  }
+
+  public send(operation: IOperation): BluebirdPromise<IRavenResponse | IRavenResponse[] | void> {
```
Comment (gregolsky) [inline src/Database/Operations/OperationExecutor.ts]: optional `sessionId` arg missing

**Comment by gregolsky** [comment]
@a-serdukov-mobilunity  There some tests failures. I presume they probably fail, since I'm running server from latest v4.0 branch. Some of them related to the etag -> change vector conversion for sure, which we have a ticket for, so these can be handled later on, but could you look through the rest of them?

```
  78 passing (30s)
  7 failing

  1) Delete command test Delete() should fail delete if etag mismatches:
     AssertionError: expected promise to be rejected but it was fulfilled with undefined


  2) Patch command test Patch request should patch success not ignoring missing:
     ConcurrencyException: Voron.Exceptions.ConcurrencyException: Could not patch document 'products/10' because non current change vector was used
   at Raven.Server.Documents.Patch.PatchDocumentCommand.Execute(DocumentsOperationContext context) in C:\work\ravendb4\src\Raven.Server\Documents\Patch\PatchDocumentCommand.cs:line 77
   at Raven.Server.Documents.TransactionOperationsMerger.ExecutePendingOperationsInTransaction(List`1 pendingOps, DocumentsOperationContext context, Task previousOperation, DurationMeasurement& meter) in C:\work\ravendb4\src\Raven.Server\Documents\TransactionOperationsMerger.cs:line 477
   at Raven.Server.Documents.TransactionOperationsMerger.MergeTransactionsOnce() in C:\work\ravendb4\src\Raven.Server\Documents\TransactionOperationsMerger.cs:line 228
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Server.Documents.TransactionOperationsMerger.<Enqueue>d__17.MoveNext() in C:\work\ravendb4\src\Raven.Server\Documents\TransactionOperationsMerger.cs:line 85
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Server.Documents.Handlers.DocumentHandler.<Patch>d__15.MoveNext() in C:\work\ravendb4\src\Raven.Server\Documents\Handlers\DocumentHandler.cs:line 488
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Server.Routing.RequestRouter.<HandlePath>d__5.MoveNext() in C:\work\ravendb4\src\Raven.Server\Routing\RequestRouter.cs:line 106
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Server.RavenServerStartup.<RequestHandler>d__12.MoveNext() in C:\work\ravendb4\src\Raven.Server\RavenServerStartup.cs:line 157
      at Function.throw (.build\src\Utility\ExceptionThrower.js:19:15)
      at Function.throwFrom (.build\src\Utility\ExceptionThrower.js:32:31)
      at Function.throwFrom (.build\src\Utility\ExceptionThrower.js:25:22)
      at PatchCommand.setResponse (.build\src\Database\RavenCommand.js:54:45)
      at PatchCommand.setResponse (.build\src\Database\Commands\PatchCommand.js:44:30)
      at RequestPromise.finally.catch.then (.build\src\Http\Request\RequestExecutor.js:161:43)
      at tryCatcher (node_modules\bluebird\js\release\util.js:16:23)
      at Promise._settlePromiseFromHandler (node_modules\bluebird\js\release\promise.js:512:31)
      at Promise._settlePromise (node_modules\bluebird\js\release\promise.js:569:18)
      at Promise._settlePromise0 (node_modules\bluebird\js\release\promise.js:614:10)
      at Promise._settlePromises (node_modules\bluebird\js\release\promise.js:693:18)
      at Async._drainQueue (node_modules\bluebird\js\release\async.js:133:16)
      at Async._drainQueues (node_modules\bluebird\js\release\async.js:143:10)
      at Immediate.Async.drainQueues (node_modules\bluebird\js\release\async.js:17:14)

  3) QueryCommand Tests "before each" hook for "should do query":
     InvalidOperationException: System.InvalidOperationException: Changing 'Products/10' from '@empty' to 'products' via update is not supported.
Delete it and recreate the document Products/10.
   at Raven.Server.Documents.DocumentPutAction.ThrowInvalidCollectionNameChange(String id, CollectionName oldCollectionName, CollectionName collectionName) in C:\work\ravendb4\src\Raven.Server\Documents\DocumentPutAction.cs:line 385
   at Raven.Server.Documents.DocumentPutAction.PutDocument(DocumentsOperationContext context, String id, String expectedChangeVector, BlittableJsonReaderObject document, Nullable`1 lastModifiedTicks, String changeVector, DocumentFlags flags, NonPersistentDocumentFlags nonPersistentFlags) in C:\work\ravendb4\src\Raven.Server\Documents\DocumentPutAction.cs:line 93
   at Raven.Server.Documents.Handlers.DocumentHandler.MergedPutCommand.Execute(DocumentsOperationContext context) in C:\work\ravendb4\src\Raven.Server\Documents\Handlers\DocumentHandler.cs:line 611
   at Raven.Server.Documents.TransactionOperationsMerger.ExecutePendingOperationsInTransaction(List`1 pendingOps, DocumentsOperationContext context, Task previousOperation, DurationMeasurement& meter) in C:\work\ravendb4\src\Raven.Server\Documents\TransactionOperationsMerger.cs:line 477
   at Raven.Server.Documents.TransactionOperationsMerger.MergeTransactionsOnce() in C:\work\ravendb4\src\Raven.Server\Documents\TransactionOperationsMerger.cs:line 228
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Server.Documents.TransactionOperationsMerger.<Enqueue>d__17.MoveNext() in C:\work\ravendb4\src\Raven.Server\Documents\TransactionOperationsMerger.cs:line 85
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Server.Documents.Handlers.DocumentHandler.<Put>d__14.MoveNext() in C:\work\ravendb4\src\Raven.Server\Documents\Handlers\DocumentHandler.cs:line 440 --- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Server.Routing.RequestRouter.<HandlePath>d__5.MoveNext() in C:\work\ravendb4\src\Raven.Server\Routing\RequestRouter.cs:line 106
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Server.RavenServerStartup.<RequestHandler>d__12.MoveNext() in C:\work\ravendb4\src\Raven.Server\RavenServerStartup.cs:line 157
      at Function.throw (.build\src\Utility\ExceptionThrower.js:19:15)
      at Function.throwFrom (.build\src\Utility\ExceptionThrower.js:32:31)
      at Function.throwFrom (.build\src\Utility\ExceptionThrower.js:25:22)
      at PutDocumentCommand.setResponse (.build\src\Database\RavenCommand.js:54:45)
      at PutDocumentCommand.setResponse (.build\src\Database\Commands\DeleteDocumentCommand.js:29:15)
      at PutDocumentCommand.setResponse (.build\src\Database\Commands\PutDocumentCommand.js:20:15)
      at RequestPromise.finally.catch.then (.build\src\Http\Request\RequestExecutor.js:161:43)
      at tryCatcher (node_modules\bluebird\js\release\util.js:16:23)
      at Promise._settlePromiseFromHandler (node_modules\bluebird\js\release\promise.js:512:31)
      at Promise._settlePromise (node_modules\bluebird\js\release\promise.js:569:18)
      at Promise._settlePromise0 (node_modules\bluebird\js\release\promise.js:614:10)
      at Promise._settlePromises (node_modules\bluebird\js\release\promise.js:693:18)
      at Async._drainQueue (node_modules\bluebird\js\release\async.js:133:16)
      at Async._drainQueues (node_modules\bluebird\js\release\async.js:143:10)
      at Immediate.Async.drainQueues (node_modules\bluebird\js\release\async.js:17:14)

  4) Document conversion test Conversion should resolve document constructors:
     AssertionError: expected [ Array(4) ] to have a length of 3 but got 4
      at Object.<anonymous> (.build\test\session_tests\DocumentConversionTest.js:89:41)
      at Generator.next (<anonymous>)
      at fulfilled (.build\test\session_tests\DocumentConversionTest.js:4:58)
      at <anonymous>

  5) Document delete test Document delete should fail delete when etag mismatches:
     AssertionError: expected promise to be rejected but it was fulfilled with undefined


  6) Document query test Index checking should query by startsWith:

      AssertionError: expected 'New Product' to equal 'new_testing'
      + expected - actual

      -New Product
      +new_testing

      at Object.<anonymous> (.build\test\session_tests\DocumentQueryTest.js:74:47)
      at Generator.next (<anonymous>)
      at fulfilled (.build\test\session_tests\DocumentQueryTest.js:4:58)
      at <anonymous>

  7) Document query test Index checking should query with ordering:
     AssertionError: expected 10 to equal 'a'
      at Object.<anonymous> (.build\test\session_tests\DocumentQueryTest.js:129:48)
      at Generator.next (<anonymous>)
      at fulfilled (.build\test\session_tests\DocumentQueryTest.js:4:58)
      at <anonymous>



npm ERR! Test failed.  See above for more details.
```

**Comment by mobilunity-user** [comment]
@gregolsky Yes there are some issues with latest server which i would to discuss this morning (i told you about them yesterday)

**Comment by gregolsky** [comment]
@ayende Let's close this one please and merge the other one #27

**Pull request closed without merging**